File: sort_functions/sort_functions_c/mergeSort.c
#include <stdio.h>


int mergeSort(int array[], int size);
int merge(int leftArray[], int leftSize, int rightArray[], int rightSize, int array[]);


int mergeSort(int array[], int size) {

    if (size <= 1) { return 0; }

    int middle = size / 2;
    int leftArray[middle];
    int rightArray[size - middle];

    int i = 0;
    int j = 0;

    for (; i < size; i++) {

        if (i < middle) {
            leftArray[i]  = array[i];
        } else {
            rightArray[j] = array[i];
            j++;
        }
    }

    int leftComparisons = mergeSort(leftArray, middle);
    int rightComparisons = mergeSort(rightArray, size-middle);

    int mergeComparisons = merge(leftArray, middle, rightArray, size-middle, array);

    return leftComparisons + rightComparisons + mergeComparisons;
}


int merge(int leftArray[], int leftSize, int rightArray[], int rightSize, int array[]) {

    int i = 0;
    int l = 0;
    int r = 0;
    int comparisons = 0;

    while (l < leftSize && r < rightSize) {

        comparisons++;

        if (leftArray[l] < rightArray[r]) {
            array[i] = leftArray[l];
            i++;
            l++;
        } else {
            array[i] = rightArray[r];
            i++;
            r++;
        }
    }

    while (l < leftSize) {
        array[i] = leftArray[l];
        i++;
        l++;
    }

    while (r < rightSize) {
        array[i] = rightArray[r];
        i++;
        r++;
    }

    return comparisons;
}

File: sort_functions/sort_functions_python/merge_sort.py
def merge_sort(arr):

    if len(arr) <= 1:
        return arr, 0

    # Split the array into two halves
    middle = len(arr) // 2
    left_half = arr[:middle]
    right_half = arr[middle:]

    # Recursively sort both halves
    left_sorted, left_comparisons = merge_sort(left_half)
    right_sorted, right_comparisons = merge_sort(right_half)

    # Merge the sorted halves and get the number of comparisons done during merge
    merged, merge_comparisons = merge(left_sorted, right_sorted)

    # Return merged array and total comparisons
    total_comparisons = left_comparisons + right_comparisons + merge_comparisons
    
    return merged, total_comparisons


def merge(left, right):

    merged = []
    left_index, right_index = 0, 0
    comparisons = 0

    # Traverse both left and right arrays
    while left_index < len(left) and right_index < len(right):

        comparisons += 1

        if left[left_index] < right[right_index]:
            merged.append(left[left_index])
            left_index += 1
        else:
            merged.append(right[right_index])
            right_index += 1

    # If left array still has elements, add them to merged list
    while left_index < len(left):
        merged.append(left[left_index])
        left_index += 1

    # If right array still has elements, add them to merged list
    while right_index < len(right):
        merged.append(right[right_index])
        right_index += 1

    return merged, comparisons